prototype/mobile_endpoint/dao.py

```diff
@@ -0,0 +1,238 @@
+from abc import ABCMeta, abstractmethod
+import functools
+import types
+from uuid import UUID
+
+import redis
+from sqlalchemy.orm import contains_eager, defer
+from mobile_endpoint.case.models import CommCareCase
+
+from mobile_endpoint.exceptions import IllegalCaseId
+from mobile_endpoint.form.models import XFormInstance
+from mobile_endpoint.models import db, Synclog, FormData, CaseData, CaseIndex
+from mobile_endpoint.utils import chunked, get_with_lock
+
+
+def to_generic(fn):
+    def _wrap(obj):
+        if hasattr(obj, 'to_generic'):
+            return obj.to_generic()
+        else:
+            return obj
+
+    @functools.wraps(fn)
+    def _inner(*args, **kwargs):
+        obj = fn(*args, **kwargs)
+        if isinstance(obj, (list, tuple)):
+            return [_wrap(ob) for ob in obj]
+        elif isinstance(obj, types.GeneratorType):
+            return (_wrap(ob) for ob in obj)
+        else:
+            return _wrap(obj)
+
+    return _inner
+
+class AbsctractDao(object):
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def commit(self, xform, cases):
+        """
+        Commit the transaction
+        """
+        pass
+
+    @abstractmethod
+    def get_synclog(self, id):
+        pass
+
+    @abstractmethod
+    def get_form(self, id):
+        pass
+    
+    @abstractmethod
+    def get_case(self, id, lock=False):
+        pass
+
+    @abstractmethod
+    def case_exists(self, id):
+        pass
+
+    @abstractmethod
+    def iter_cases(self, case_ids, chunksize=100, ordered=True):
+        pass
+
+    @abstractmethod
+    def get_reverse_indexed_cases(self, domain, case_ids):
```

Comment:
i was surprised not to see get_indexed_cases also in here. i think that will end up being necessary for restores